### The pandas library

__Popular Python module that provides tools for data analysis.__

> Pandas is widely used to transform raw data for data analysis and machine learning. 

We will learn: 

- data frames and data series 
- reading from files 
- data transformation 
- statistical analysis


#### The Pandas library two types: 

- __Series__

     Handles pairs of values, where one value is the index. 
    
     Example 'date' and 'sales for the day'.

- __Dataframe__

    Handles tables with multiple columns. 


In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import pprint
printer_obj = pprint.PrettyPrinter(indent=4)
%matplotlib inline

In [2]:
# Helper functions

import inspect

def print_name_value(variable):
    frame = inspect.currentframe()
    frame = inspect.getouterframes(frame)[1]
    ctx = inspect.getframeinfo(frame[0]).code_context[0].strip()
    single_arg = ctx[ctx.find('(') + 1:-1].split(',')[0]
    mem_variable = id(variable)
    print(f'{single_arg}:\n{variable}\n')
 
def print_name_type(data_obj):
    frame = inspect.currentframe()
    frame = inspect.getouterframes(frame)[1]
    ctx = inspect.getframeinfo(frame[0]).code_context[0].strip()
    var_name = ctx[ctx.find('(') + 1:-1].split(',')[0]
    obj_type = get_pandas_type(data_obj)
    print(f'The variable "{var_name}" is a {obj_type}')

def print_header(title='', sep='.', witdh=46):
    '''Prints a formated header.'''
    title_size = len(title)
    filler_size = (witdh - (title_size + 2)) // 2
    filler_str = sep * filler_size
    if len(title) == 0:
        rfill = filler_str
    else: 
        rfill = filler_str[::-1]
        if (title_size % 2 == 1):
             rfill = rfill + rfill[0]
    print('{left}[{title}]{right}'.format(
        left=filler_str, title=title, right=rfill))

def print_separator():
    print('. ' * 25)
    
def pp(obj):
    printer_obj.pprint(obj)

def get_pandas_type(data_obj):
    type_str = str(type(data_obj))
    return type_str.split('.')[-1].split("'")[0]

def print_type(data_obj):
    print(get_pandas_type(data_obj))
    
def print_data_and_type(data_obj, title:str=None):
    if title:
        print_header(title, sep='-')
        print(' ')
    print('Data:')
    if isinstance(data_obj, pd.Series):
        lines = str(data_obj).split('\n')[:-1]
        print('\n'.join(lines))
    else:
        print(data_obj[:])
    print(' ')
    print('Object Type: {0}'.format(get_pandas_type(data_obj)))
    print(' ')

def print_shape(df):
    print('Shape: ' + str(df.shape))
    
def print_columns(df):
    print('-- Columns --')
    for col in df.columns:
        print(' ' + col)

#### Creating a data series from a list

In [3]:
list1 = ['Grapes', 'Apples', 'Oranges', 'Bananas']
s1 = pd.Series(list1)
print_data_and_type(s1)

Data:
0     Grapes
1     Apples
2    Oranges
3    Bananas
 
Object Type: Series
 


Notice that rows are given numbers, these numbers are known as indexes. 

By default, indexes start from 0 and go up.  

#### Using custom indexes 

In [4]:
# defining list2 with indices
my_index = ['GR', 'AP', 'OR', 'BA']

# In series1 we say index=my_index
series1 = pd.Series(list1, index=my_index)
print_data_and_type(series1)

Data:
GR     Grapes
AP     Apples
OR    Oranges
BA    Bananas
 
Object Type: Series
 


#### Creating a series from a python dictionary.

In [5]:
dict1 = {'Z': 'Zynga', 'U': 'Uber', 'G':'Google'}
ds1 = pd.Series(dict1)
print_data_and_type(ds1)

Data:
Z     Zynga
U      Uber
G    Google
 
Object Type: Series
 


#### Pandas internally uses a library called Numpy. 
(More on this in future sessions.)

In [6]:
# Generating some arrays with Numpy
arr = np.arange(10, 15)
print(arr)
print(arr[4])
print(np.mean(arr))

[10 11 12 13 14]
14
12.0


In [7]:
# There are other libraries that operate directly on python lists.
import statistics
print_header('Array')
a = [x for x in range(1,10)]
print(a)
print_header('Mean')
statistics.mean(a)

...................[Array]....................
[1, 2, 3, 4, 5, 6, 7, 8, 9]
....................[Mean]....................


5

In [8]:
# Nympy has an equivalent type called 'array'
b = np.arange(20, 30)
print_header('Array')
print(b)
# transform into a 2x5 matrix.
c = b.reshape(2, 5)
print_header('Matrix')
print(c)

...................[Array]....................
[20 21 22 23 24 25 26 27 28 29]
...................[Matrix]...................
[[20 21 22 23 24]
 [25 26 27 28 29]]


#### Creating a Series with 'company name' as index and its stock 'price' as values. 

##### Notice that in the code below, we have a dictionary with two keys having 'None' as their values.

In [9]:
d2 = {
    'Amazon': 3352, 'Nvidia': None, 'Alphabet': 856, 'Toyota': '112', 
    'GE': 29, 'Ford': 12, 'Marriot': None, 'amazon': 1000
}
print_header('Dictionary', sep='-')
pp(d2)
companies_incomplete = pd.Series(d2, name='Price')
print_data_and_type(companies_incomplete, 'Pandas Series')

-----------------[Dictionary]-----------------
{   'Alphabet': 856,
    'Amazon': 3352,
    'Ford': 12,
    'GE': 29,
    'Marriot': None,
    'Nvidia': None,
    'Toyota': '112',
    'amazon': 1000}
---------------[Pandas Series]----------------
 
Data:
Amazon      3352
Nvidia      None
Alphabet     856
Toyota       112
GE            29
Ford          12
Marriot     None
amazon      1000
 
Object Type: Series
 


In [10]:
# Lets use a dictionary with all values filled for now.
d2 = {
    'Amazon': 3352, 'Nvidia': 22, 'Alphabet': 856, 'Toyota': 112, \
    'GE': 29, 'Ford': 12, 'Marriot': 33, 'amazon': 1000}
print_header('Dictionary', sep='-')
pp(d2)
companies = pd.Series(d2, name='Price')
print(' ')
print_data_and_type(companies, 'Pandas Series')

-----------------[Dictionary]-----------------
{   'Alphabet': 856,
    'Amazon': 3352,
    'Ford': 12,
    'GE': 29,
    'Marriot': 33,
    'Nvidia': 22,
    'Toyota': 112,
    'amazon': 1000}
 
---------------[Pandas Series]----------------
 
Data:
Amazon      3352
Nvidia        22
Alphabet     856
Toyota       112
GE            29
Ford          12
Marriot       33
amazon      1000
 
Object Type: Series
 


Using the index we can get the corresponding stock prices. Below we are 
retrieving the stock price of Ford.

In [11]:
print(companies[['Ford']])

Ford    12
Name: Price, dtype: int64


We can also get stock prices for more companies. We have to supply the indices 
that we are interested in as a list.

In [12]:
print(companies[['Ford', 'GE']])

Ford    12
GE      29
Name: Price, dtype: int64


Membership can be checked using the 'in' keyword.

In [13]:
print('Amazon' in companies)
print('Apple' in companies)

True
False


If we want to know companies for which we don't have stock price, then we have 
to use the is.null(). True will be returned for the indices that don't have a 
value and False for the indices that have 
a value.

In [14]:
print(companies_incomplete.isnull())

Amazon      False
Nvidia       True
Alphabet    False
Toyota      False
GE          False
Ford        False
Marriot      True
amazon      False
Name: Price, dtype: bool


In pandas if you have column of boolean, you can use this info to select records on the same data.

In [15]:
# Printing only the companies where price is null.
print(companies_incomplete[companies_incomplete.isnull()])

Nvidia     None
Marriot    None
Name: Price, dtype: object


In [16]:
# Use the tilde to invert the logic
# Printing only the companies where price is not null.
print_data_and_type(companies_incomplete[~companies_incomplete.isnull()])

Data:
Amazon      3352
Alphabet     856
Toyota       112
GE            29
Ford          12
amazon      1000
 
Object Type: Series
 


#### Dataframes

A dataframe is a tabular data structure that consists of rows and columns. 
Dataframe is nothing but a collection of series.

##### Let's create a dataframe using a dicitonary as shown below.

In [17]:
c1 = {
    'Name': ['Amazon', 'GE', 'Toyota', 'Twitter', 'Ford', 'Marriot'],
    'Founded': [1994, 1923, 1937, 2006, 1903, 1927], 
    'Price': [852, 111.2, 112, 15.2, 12.5, 88.31],
    'Unused': [1, 2, 3, 4, 5, 6]
}

companies = pd.DataFrame(c1, columns=['Name', 'Founded', 'Price'])
print_data_and_type(companies)

Data:
      Name  Founded   Price
0   Amazon     1994  852.00
1       GE     1923  111.20
2   Toyota     1937  112.00
3  Twitter     2006   15.20
4     Ford     1903   12.50
5  Marriot     1927   88.31
 
Object Type: DataFrame
 


In [18]:
c1 = {
    'Name': ['Amazon', 'GE', 'Toyota', 'Twitter', 'Ford', 'Marriot'],
    'Founded': [1994, 1923, 1937, 2006, 1903, 1927], 
    'Price': [852, 111.2, 112, 15.2, 12.5, 88.31],
    'Unused_1': [ 1, 2, 3, 4, 5, 6],
    'Unused_2': [11,12,13,14,15,16]
}

companies_raw = pd.DataFrame(c1)
print_data_and_type(companies_raw, 'Raw')
companies_clean = companies_raw.drop(['Unused_1','Unused_2'], axis=1)
print_data_and_type(companies_clean, 'Clean')


--------------------[Raw]---------------------
 
Data:
      Name  Founded   Price  Unused_1  Unused_2
0   Amazon     1994  852.00         1        11
1       GE     1923  111.20         2        12
2   Toyota     1937  112.00         3        13
3  Twitter     2006   15.20         4        14
4     Ford     1903   12.50         5        15
5  Marriot     1927   88.31         6        16
 
Object Type: DataFrame
 
-------------------[Clean]--------------------
 
Data:
      Name  Founded   Price
0   Amazon     1994  852.00
1       GE     1923  111.20
2   Toyota     1937  112.00
3  Twitter     2006   15.20
4     Ford     1903   12.50
5  Marriot     1927   88.31
 
Object Type: DataFrame
 


In [19]:
capitals = ["Sacramento", "Boston", "Austin", "Denver", "Phoenix"]
capseries = pd.Series(capitals)
print_data_and_type(capseries, 'Capitals')


------------------[Capitals]------------------
 
Data:
0    Sacramento
1        Boston
2        Austin
3        Denver
4       Phoenix
 
Object Type: Series
 


In [20]:
state = ["CA", "MA", "TX", "CO", "AZ"]
capstate = pd.Series(capitals, index=state)
print_data_and_type(capstate)


Data:
CA    Sacramento
MA        Boston
TX        Austin
CO        Denver
AZ       Phoenix
 
Object Type: Series
 


In [21]:
capital_list = ['Springfield','Austin','Albany','Sacramento','WHAT','WHAT_1']
list_1 = pd.Series(capital_list)
print_data_and_type(list_1, 'Std Index')
print(' ')
# defining idx with indices
idx = ["IL","TX","NY","CA", "W1","W2"]
s_2 = pd.Series(capital_list, index = idx)
print_data_and_type(s_2, 'Custom Index')

-----------------[Std Index]------------------
 
Data:
0    Springfield
1         Austin
2         Albany
3     Sacramento
4           WHAT
5         WHAT_1
 
Object Type: Series
 
 
----------------[Custom Index]----------------
 
Data:
IL    Springfield
TX         Austin
NY         Albany
CA     Sacramento
W1           WHAT
W2         WHAT_1
 
Object Type: Series
 


## Reading files

Using the dataset from: https://www.kaggle.com/deepmatrix/imdb-5000-movie-dataset

Reading a csv file. 

In [22]:
movies = pd.read_csv('data/imdb_movie/movie_metadata.csv')
movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [23]:
print_name_type(movies)

The variable "movies" is a DataFrame


In [24]:
# Getting the list of columns and printing them manually.
# We have a helper function for that.
print_columns(movies)
#
# Or you can use the code below:
#
# column_names = movies.columns
# for column in column_names:
#     print(column)

-- Columns --
 color
 director_name
 num_critic_for_reviews
 duration
 director_facebook_likes
 actor_3_facebook_likes
 actor_2_name
 actor_1_facebook_likes
 gross
 genres
 actor_1_name
 movie_title
 num_voted_users
 cast_total_facebook_likes
 actor_3_name
 facenumber_in_poster
 plot_keywords
 movie_imdb_link
 num_user_for_reviews
 language
 country
 content_rating
 budget
 title_year
 actor_2_facebook_likes
 imdb_score
 aspect_ratio
 movie_facebook_likes


In [25]:
# You can use the head function from Dataframes to show top X records
movies_gr = movies[['aspect_ratio']]
print_name_type(movies_gr)
movies_gr.head(8)

The variable "movies_gr" is a DataFrame


aspect_ratio
0          1.78
1          2.35
2          2.35
3          2.35
4           NaN
5          2.35
6          2.35
7          1.85

In [26]:
# Group and count
counts = movies_gr.groupby('aspect_ratio').size()
print_data_and_type(counts, 'Counts grouped')
framed_counts = counts.to_frame()
print_data_and_type(framed_counts, 'Framed counts')

---------------[Counts grouped]---------------
 
Data:
aspect_ratio
1.18        1
1.20        1
1.33       68
1.37      100
1.44        1
1.50        2
1.66       64
1.75        3
1.77        1
1.78      110
1.85     1906
1.89        1
2.00        5
2.20       15
2.24        1
2.35     2360
2.39       15
2.40        3
2.55        2
2.76        3
4.00        7
16.00      45
 
Object Type: Series
 
---------------[Framed counts]----------------
 
Data:
                 0
aspect_ratio      
1.18             1
1.20             1
1.33            68
1.37           100
1.44             1
1.50             2
1.66            64
1.75             3
1.77             1
1.78           110
1.85          1906
1.89             1
2.00             5
2.20            15
2.24             1
2.35          2360
2.39            15
2.40             3
2.55             2
2.76             3
4.00             7
16.00           45
 
Object Type: DataFrame
 


In [113]:
# Doing the grouping and sorting at the same time.
counts = movies_gr.groupby('aspect_ratio').size().sort_values(ascending=False)
# counts.reset_index(inplace=True, drop=True)
# counts.reset_index(inplace=True)
type(counts)

pandas.core.series.Series

In [126]:
# Converting to Dataframe
counts_df = counts.to_frame()
counts_df.reset_index(inplace=True)
counts_df.columns = ['Aspect Ratio', 'Counts'] 
counts_df.head()

Aspect Ratio  Counts
0          2.35    2360
1          1.85    1906
2          1.78     110
3          1.37     100
4          1.33      68

In [123]:
type(counts_df)

pandas.core.frame.DataFrame

In [131]:
# The 'shape' field contains a tuple with (num_rows, num_cols).
num_rows, num_cols = movies.shape
print('Rows : {0:>4}'.format(num_rows))
print('Cols : {0:>4}'.format(num_cols))

Rows : 5043
Cols :   28


We can create a series from the movies dataframe. 

In [130]:
movies_dir = movies['director_name']
print_data_and_type(movies_dir)

Data:
0           James Cameron
1          Gore Verbinski
2              Sam Mendes
3       Christopher Nolan
4             Doug Walker
              ...        
5038          Scott Smith
5039                  NaN
5040     Benjamin Roberds
5041          Daniel Hsia
5042             Jon Gunn
 
Object Type: Series
 


#### A derived dataframe with columns: movie_title, duration, budget, gross, genres, director_name.

In [32]:
slim_movies_raw = movies[
    ['movie_title', 'duration', 'budget', 'gross', 
     'genres', 'director_name']]
slim_movies_raw.head(3)

movie_title  duration       budget  \
0                                    Avatar      178.0  237000000.0   
1  Pirates of the Caribbean: At World's End      169.0  300000000.0   
2                                   Spectre      148.0  245000000.0   

         gross                           genres   director_name  
0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   James Cameron  
1  309404152.0         Action|Adventure|Fantasy  Gore Verbinski  
2  200074175.0        Action|Adventure|Thriller      Sam Mendes

In [33]:
slim_movies_raw.shape

(5043, 6)

We can separate the genre string into an array of values

In [34]:
genre_series = slim_movies_raw['genres'].str.split('|')
slim_movies = slim_movies_raw.copy()
print('genre_series={0}'.format(genre_series.shape))
slim_movies = slim_movies.assign(genres=genre_series)
slim_movies.head(3)

genre_series=(5043,)


movie_title  duration       budget  \
0                                    Avatar      178.0  237000000.0   
1  Pirates of the Caribbean: At World's End      169.0  300000000.0   
2                                   Spectre      148.0  245000000.0   

         gross                                genres   director_name  
0  760505847.0  [Action, Adventure, Fantasy, Sci-Fi]   James Cameron  
1  309404152.0          [Action, Adventure, Fantasy]  Gore Verbinski  
2  200074175.0         [Action, Adventure, Thriller]      Sam Mendes

We can use sort_values() to sort a dataframe.

In [132]:
sorted_movies = slim_movies.sort_values('movie_title')
sorted_movies.head(5)

movie_title  duration      budget       gross  \
4447                     #Horror      101.0   1500000.0         NaN   
3698         10 Cloverfield Lane      104.0  15000000.0  71897215.0   
3015       10 Days in a Madhouse      111.0  12000000.0     14616.0   
2845  10 Things I Hate About You       97.0  16000000.0  38176108.0   
279      10,000 B.C.                   22.0         NaN         NaN   

                                          genres        director_name  
4447          [Drama, Horror, Mystery, Thriller]         Tara Subkoff  
3698  [Drama, Horror, Mystery, Sci-Fi, Thriller]     Dan Trachtenberg  
3015                                     [Drama]        Timothy Hines  
2845                    [Comedy, Drama, Romance]           Gil Junger  
279                                     [Comedy]  Christopher Barnard

Notice that not all numeric values in our data frame are valid. 

Pands represents these values with a special constant 'NaN' that means 'Not a number'

We want to drop rows that have a NaN in any column. 

In [72]:
movies_no_nulls = slim_movies.copy(deep=True)
movies_no_nulls.dropna(how='any', inplace=True)
print_shape(movies_no_nulls)
movies_no_nulls.head()

Shape: (3890, 6)


movie_title  duration       budget  \
0                                    Avatar      178.0  237000000.0   
1  Pirates of the Caribbean: At World's End      169.0  300000000.0   
2                                   Spectre      148.0  245000000.0   
3                     The Dark Knight Rises      164.0  250000000.0   
5                               John Carter      132.0  263700000.0   

         gross                                genres      director_name  
0  760505847.0  [Action, Adventure, Fantasy, Sci-Fi]      James Cameron  
1  309404152.0          [Action, Adventure, Fantasy]     Gore Verbinski  
2  200074175.0         [Action, Adventure, Thriller]         Sam Mendes  
3  448130642.0                    [Action, Thriller]  Christopher Nolan  
5   73058679.0           [Action, Adventure, Sci-Fi]     Andrew Stanton

Notice the number of rows now is smaller (3890 instead of 5043)


In [36]:
# When making modifications in dataframes sometimes we want to keep the original version.
# so we use the 'copy' call to make a duplicate of the data.
mov_with_budget_duration = slim_movies.copy(deep=True)
mov_with_budget_duration.dropna(subset=['duration','budget'], how='any', inplace=True)
print_shape(mov_with_budget_duration)
mov_with_budget_duration.head()

Shape: (4546, 6)


movie_title  duration       budget  \
0                                    Avatar      178.0  237000000.0   
1  Pirates of the Caribbean: At World's End      169.0  300000000.0   
2                                   Spectre      148.0  245000000.0   
3                     The Dark Knight Rises      164.0  250000000.0   
5                               John Carter      132.0  263700000.0   

         gross                           genres      director_name  
0  760505847.0  Action|Adventure|Fantasy|Sci-Fi      James Cameron  
1  309404152.0         Action|Adventure|Fantasy     Gore Verbinski  
2  200074175.0        Action|Adventure|Thriller         Sam Mendes  
3  448130642.0                  Action|Thriller  Christopher Nolan  
5   73058679.0          Action|Adventure|Sci-Fi     Andrew Stanton

#### Filtering the Data using a boolean expression

In [37]:
# Build a filter boolean array
# Inside of the pandas the logical operations are defined as follows:
# & = and,  | = or,  ^ = xor,  ~ = not
filter_1 = ( (mov_with_budget_duration['gross']>350_000) & 
             (mov_with_budget_duration['duration']<150) )
filter_1.value_counts()

True     3391
False    1155
dtype: int64

In [38]:
filter_2 = (
    (mov_with_budget_duration['gross']>350_000) | 
    (mov_with_budget_duration['duration']<150)
)
filter_2.value_counts()

True     4516
False      30
dtype: int64

In [39]:
filter_2 = (
    (mov_with_budget_duration['gross']>350_000) ^ 
    (mov_with_budget_duration['duration']<150)
)
filter_2.value_counts()

False    3421
True     1125
dtype: int64

In [40]:
# The inverse of a and b
# Noticed the results from filter_1
filter_3 = ~ ( (mov_with_budget_duration['gross']>350_000) & 
               (mov_with_budget_duration['duration']<150) )
filter_3.value_counts()

False    3391
True     1155
dtype: int64

In [41]:
# Applies the filter to the data.
filtered_df = mov_with_budget_duration[filter_1]
print_shape(filtered_df)

Shape: (3391, 6)


In [43]:
# Select only movies with large budgets
large_gross = mov_with_budget_duration[mov_with_budget_duration['gross']>350_000]
print_shape(large_gross)
large_gross.head()

Shape: (3570, 6)


movie_title  duration       budget  \
0                                    Avatar      178.0  237000000.0   
1  Pirates of the Caribbean: At World's End      169.0  300000000.0   
2                                   Spectre      148.0  245000000.0   
3                     The Dark Knight Rises      164.0  250000000.0   
5                               John Carter      132.0  263700000.0   

         gross                           genres      director_name  
0  760505847.0  Action|Adventure|Fantasy|Sci-Fi      James Cameron  
1  309404152.0         Action|Adventure|Fantasy     Gore Verbinski  
2  200074175.0        Action|Adventure|Thriller         Sam Mendes  
3  448130642.0                  Action|Thriller  Christopher Nolan  
5   73058679.0          Action|Adventure|Sci-Fi     Andrew Stanton

Lets check the counts of null values in the dataframe we are using.

In [133]:
sorted_movies.isnull().sum()

movie_title        0
duration          15
budget           492
gross            884
genres             0
director_name    104
dtype: int64

Lets drop all rows with any nulls.

For the __gross__ values we can set the missing values to the mean.

In [134]:
# Using copy to avoid warnings.
movie_copy = sorted_movies.copy()
mean_gross = sorted_movies["gross"].mean()
print(f'Mean Gross Amt = {mean_gross:,.2f}')


Mean Gross Amt = 48,468,407.53


In [77]:
movie_copy["gross"].fillna(value=mean_gross, inplace=True)
print(movie_copy["gross"].isnull().sum())

0


For the __duration__, we can set the missing values to 90.

In [79]:
movie_copy['duration'].fillna(value=90, inplace=True)

In [80]:
movie_copy.isnull().sum()

movie_title        0
duration           0
budget           492
gross              0
genres             0
director_name    104
dtype: int64

In [135]:
clean_movies = movie_copy.copy().dropna() 

All values for the fields of interest are filled.

In [137]:
clean_movies.isnull().sum()

movie_title      0
duration         0
budget           0
gross            0
genres           0
director_name    0
dtype: int64

In [138]:
clean_movies.describe()

duration        budget         gross
count  3890.000000  3.890000e+03  3.890000e+03
mean    109.896401  4.520391e+07  5.106809e+07
std      22.706775  2.224177e+08  6.980668e+07
min      34.000000  2.180000e+02  1.620000e+02
25%      95.000000  1.000000e+07  6.844452e+06
50%     106.000000  2.400000e+07  2.799697e+07
75%     120.000000  5.000000e+07  6.540649e+07
max     330.000000  1.221550e+10  7.605058e+08

In [139]:
clean_movies.head()

movie_title  duration      budget       gross  \
3698         10 Cloverfield Lane      104.0  15000000.0  71897215.0   
3015       10 Days in a Madhouse      111.0  12000000.0     14616.0   
2845  10 Things I Hate About You       97.0  16000000.0  38176108.0   
406               102 Dalmatians      100.0  85000000.0  66941559.0   
3420                 10th & Wolf      107.0   8000000.0     53481.0   

                                          genres     director_name  
3698  [Drama, Horror, Mystery, Sci-Fi, Thriller]  Dan Trachtenberg  
3015                                     [Drama]     Timothy Hines  
2845                    [Comedy, Drama, Romance]        Gil Junger  
406                  [Adventure, Comedy, Family]        Kevin Lima  
3420                    [Crime, Drama, Thriller]    Robert Moresco

Let's create another field with the __budget in millions__.

In [154]:
movie_exp = clean_movies.copy()
movie_exp['budget_in_mm'] = movie_exp['budget'].apply(lambda x:x/1_000_000)
movie_exp[['movie_title','budget_in_mm']].head()


movie_title  budget_in_mm
3698         10 Cloverfield Lane           15.0
3015       10 Days in a Madhouse           12.0
2845  10 Things I Hate About You           16.0
406               102 Dalmatians           85.0
3420                 10th & Wolf            8.0

In [156]:
# Combining two or more columns to produce new columns
pd.set_option('display.precision', 3)
movie_test1 = movie_exp.copy()
movie_test1['profit_in_mm'] = (movie_test1['gross']-movie_test1['budget'])/1_000_000
movie_test1[['movie_title', 'gross', 'budget', 'profit_in_mm']].head()


movie_title      gross     budget  profit_in_mm
3698         10 Cloverfield Lane   7.190e+07  1.500e+07        56.897
3015       10 Days in a Madhouse   1.462e+04  1.200e+07       -11.985
2845  10 Things I Hate About You   3.818e+07  1.600e+07        22.176
406               102 Dalmatians   6.694e+07  8.500e+07       -18.058
3420                 10th & Wolf   5.348e+04  8.000e+06        -7.947

In [157]:
# Combining two or more columns using assign
movie_test2 = movie_exp.copy()
movie_test2 = movie_test2.assign(
    profit_in_mm=(movie_test2.gross - movie_test2.budget)/1_000_000)
movie_test2[['movie_title', 'gross', 'budget', 'profit_in_mm']].head()

movie_title      gross     budget  profit_in_mm
3698         10 Cloverfield Lane   7.190e+07  1.500e+07        56.897
3015       10 Days in a Madhouse   1.462e+04  1.200e+07       -11.985
2845  10 Things I Hate About You   3.818e+07  1.600e+07        22.176
406               102 Dalmatians   6.694e+07  8.500e+07       -18.058
3420                 10th & Wolf   5.348e+04  8.000e+06        -7.947

Selecting the fields that I need.

In [158]:
movie_df_raw = movie_exp.copy()
movie_df_raw = movie_df_raw.assign(
    profit_in_mm=(movie_df_raw.gross - movie_df_raw.budget)/1_000_000,
    budget_in_mm=movie_df_raw.budget/1_000_000,
    gross_in_mm=movie_df_raw.gross/1_000_000)
 
movie_df = movie_df_raw[['movie_title', 'gross_in_mm', 'budget_in_mm', 'profit_in_mm']]
movie_df.head()

movie_title  gross_in_mm  budget_in_mm  profit_in_mm
3698         10 Cloverfield Lane        71.897          15.0        56.897
3015       10 Days in a Madhouse         0.015          12.0       -11.985
2845  10 Things I Hate About You        38.176          16.0        22.176
406               102 Dalmatians        66.942          85.0       -18.058
3420                 10th & Wolf         0.053           8.0        -7.947

### Reading files from the internet.

In [166]:
from os.path import exists

def cached_load_df(url, local_file, use_cache=True):
    if use_cache and exists(local_file):
        print('Loading from cache,', end=' ')
        result = pd.read_csv(local_file)
    else:
        result = pd.read_csv(url)
        result.to_csv(local_file, index=False)
        print('Saving to cache,', end=' ')
    print('Loaded {0} records'.format(result.shape[0]))
    return result


def load_countries_df(use_cache=True):
    return cached_load_df(
        url= 'https://mirror-v01.s3.us-west-2.amazonaws.com/countries.csv', 
        local_file = 'my_countries.csv', 
        use_cache=use_cache)

countries = load_countries_df(use_cache=False)

Saving to cache, Loaded 194 records


In [167]:
countries2 = load_countries_df(use_cache=True)

Loading from cache, Loaded 194 records


In [168]:
print(countries.head())
print(countries.shape)
print('Num rows = ', end=' ')
print(countries.shape[0])

    Country  Region
0   Algeria  AFRICA
1    Angola  AFRICA
2     Benin  AFRICA
3  Botswana  AFRICA
4   Burkina  AFRICA
(194, 2)
Num rows =  194


In [169]:
print(countries.dtypes)

Country    object
Region     object
dtype: object


In [170]:
print(countries.describe())

        Country  Region
count       194     194
unique      194       6
top     Algeria  AFRICA
freq          1      54


In [171]:
company = pd.read_csv('company.csv')
company = company.set_index('Name')
company.head()

sales_budget  marketing_budget  Income
Name                                        
PLP            123                90     400
MX             340                85     700
YP              45                55     210
UP              27                60     200
QT              80               100     225

Creating a series with column sales_budget.

##### Generate Series from a DataFrame 

In [173]:
# Code Style Suggestions 

budget = 200_000

print('Wrong way')
if budget > 170_000:
    print('  large')
    
print('Right way')
threshold_for_large_budget = 170_000 # We looked at the data picket top 80 percentile.
 
if budget > threshold_for_large_budget:
    print('  large')

 

Wrong way
  large
Right way
  large


#### Dropping columns and selecting columns

In [174]:
test1 = {
    'Name': ['Amazon', 'GE', 'Toyota', 'Twitter', 'Ford', 'Marriot'],
    'Founded': [1994, 1923, 1937, 2006, 1903, 1927]}

df1 = pd.DataFrame(test1)
print_header('Original', sep=' ')
print_data_and_type(df1)
print(' ')

df2 = df1.drop(['Founded'], axis=1)
print_header('After Drop ', sep=' ')
print_data_and_type(df2)
print(' ')

print_header('After selecting one column', sep=' ')
ser1 = df1['Name']
print_data_and_type(ser1)
print(' ')

print_header('After selecting one column with list of fields', sep=' ')
ser1 = df1[['Name']]
print_data_and_type(ser1)
print(' ')


                  [Original]                  
Data:
      Name  Founded
0   Amazon     1994
1       GE     1923
2   Toyota     1937
3  Twitter     2006
4     Ford     1903
5  Marriot     1927
 
Object Type: DataFrame
 
 
                [After Drop ]                 
Data:
      Name
0   Amazon
1       GE
2   Toyota
3  Twitter
4     Ford
5  Marriot
 
Object Type: DataFrame
 
 
         [After selecting one column]         
Data:
0     Amazon
1         GE
2     Toyota
3    Twitter
4       Ford
5    Marriot
 
Object Type: Series
 
 
[After selecting one column with list of fields]
Data:
      Name
0   Amazon
1       GE
2   Toyota
3  Twitter
4     Ford
5  Marriot
 
Object Type: DataFrame
 
 


#### Pandas slicing with iLoc vs Loc

In [175]:
company.head(10)

sales_budget  marketing_budget  Income
Name                                        
PLP            123                90     400
MX             340                85     700
YP              45                55     210
UP              27                60     200
QT              80               100     225
AD              95               100     300
XC              26                40     150
MA              72                90     220
GP              66                80     290
LA              54                70     240

In [176]:
# Selecting row & columns using integer references.
company.iloc[3:5, 0:2]

sales_budget  marketing_budget
Name                                
UP              27                60
QT              80               100

In [177]:
# Selecting row & columns using names.
company.loc['PLP':'QT', ['sales_budget', 'marketing_budget']]

sales_budget  marketing_budget
Name                                
PLP            123                90
MX             340                85
YP              45                55
UP              27                60
QT              80               100

In [178]:
company_orig = pd.read_csv('company.csv')
print_data_and_type(company_orig)

Data:
   Name  sales_budget  marketing_budget  Income
0   PLP           123                90     400
1    MX           340                85     700
2    YP            45                55     210
3    UP            27                60     200
4    QT            80               100     225
5    AD            95               100     300
6    XC            26                40     150
7    MA            72                90     220
8    GP            66                80     290
9    LA            54                70     240
10   WT            43                50     150
 
Object Type: DataFrame
 


In [179]:
company_orig.loc[4:6, ['sales_budget', 'marketing_budget']]

sales_budget  marketing_budget
4            80               100
5            95               100
6            26                40

#### Creating new fields

In [180]:
company_new = pd.read_csv('company.csv')
print_data_and_type(company_new)

Data:
   Name  sales_budget  marketing_budget  Income
0   PLP           123                90     400
1    MX           340                85     700
2    YP            45                55     210
3    UP            27                60     200
4    QT            80               100     225
5    AD            95               100     300
6    XC            26                40     150
7    MA            72                90     220
8    GP            66                80     290
9    LA            54                70     240
10   WT            43                50     150
 
Object Type: DataFrame
 


In [181]:
company_new['large_budget'] = company_new.sales_budget > 100
company_new.head(4)

Name  sales_budget  marketing_budget  Income  large_budget
0  PLP           123                90     400          True
1   MX           340                85     700          True
2   YP            45                55     210         False
3   UP            27                60     200         False

#### Assigning with lambda

In [182]:
company_new = pd.read_csv('company.csv')
company_with_lambda = company_new.assign(
    large_sales_budget=company_new.sales_budget > 100,
    large_marketing_budget=company_new.marketing_budget > 40,
    # Defining a lambda function here
    both_large=lambda d: d.large_sales_budget & d.large_marketing_budget
    )
company_with_lambda.head(10)

Name  sales_budget  marketing_budget  Income  large_sales_budget  \
0  PLP           123                90     400                True   
1   MX           340                85     700                True   
2   YP            45                55     210               False   
3   UP            27                60     200               False   
4   QT            80               100     225               False   
5   AD            95               100     300               False   
6   XC            26                40     150               False   
7   MA            72                90     220               False   
8   GP            66                80     290               False   
9   LA            54                70     240               False   

   large_marketing_budget  both_large  
0                    True        True  
1                    True        True  
2                    True       False  
3                    True       False  
4                    True       False  
5                    True       False  
6                   False       False  
7                    True       False  
8                    True       False  
9                    True       False

#### Wide and Long Datasets

In [186]:
data = pd.DataFrame({
    'Day':   ['08/01', '08/02', '08/03', '08/04',    '08/05'],
    'Ford':   [1.03,    1.05,    1.07,    1.09,      1.11],
    'Apple':  [2.01,    2.03,    2.05,    2.07,      2.09],
    'Amazon': [3.02,    3.04,    3.06,    3.08,      3.10],
    'GE':     [4.04,    4.05,    4.06,    4.07,      4.08],
})


#### Converting to long with 'melt'

In [191]:
melted_data = data.melt(id_vars=['Day'], var_name='Company', value_name='Price')
melted_data.head(10)

Day Company  Price
0  08/01    Ford   1.03
1  08/02    Ford   1.05
2  08/03    Ford   1.07
3  08/04    Ford   1.09
4  08/05    Ford   1.11
5  08/01   Apple   2.01
6  08/02   Apple   2.03
7  08/03   Apple   2.05
8  08/04   Apple   2.07
9  08/05   Apple   2.09

#### Converting back to wide with 'pivot'

In [192]:
# Convert from long to wide
pivoted_data = melted_data.pivot(index="Day", columns=["Company"], values="Price")
pivoted_data = pivoted_data.reset_index(inplace=False)# .rename_axis(None,1)
pivoted_data.columns.name = ' ' 
pivoted_data

Day  Amazon  Apple  Ford    GE
0  08/01    3.02   2.01  1.03  4.04
1  08/02    3.04   2.03  1.05  4.05
2  08/03    3.06   2.05  1.07  4.06
3  08/04    3.08   2.07  1.09  4.07
4  08/05    3.10   2.09  1.11  4.08

#### Back to long with 'melt'

In [195]:
back_to_melted  = pivoted_data.melt(id_vars=['Day'])
back_to_melted.columns = ['Day', 'Company', 'Price']
back_to_melted.head(10)

Day Company  Price
0  08/01  Amazon   3.02
1  08/02  Amazon   3.04
2  08/03  Amazon   3.06
3  08/04  Amazon   3.08
4  08/05  Amazon   3.10
5  08/01   Apple   2.01
6  08/02   Apple   2.03
7  08/03   Apple   2.05
8  08/04   Apple   2.07
9  08/05   Apple   2.09

#### Using map and 